libclc/ptx-nvidiacl/libspirv/images/image.cl

```diff
@@ -256,36 +256,47 @@ pixelf32 as_pixelf32(int4 v) { return as_float4(v); }
             image, x * sizeof(pixelf##pixelf_size), y, z));                    \
   }
 
-_DEFINE_VEC4_CAST(float, int)
-_DEFINE_VEC4_CAST(int, float)
-_DEFINE_VEC4_CAST(float, uint)
-_DEFINE_VEC4_CAST(uint, float)
-_DEFINE_VEC4_CAST(uint, int)
+_DEFINE_VEC4_CAST(int, int)
 _DEFINE_VEC4_CAST(int, uint)
 _DEFINE_VEC4_CAST(int, short)
 _DEFINE_VEC4_CAST(int, char)
+_DEFINE_VEC4_CAST(int, float)
 _DEFINE_VEC4_CAST(uint, ushort)
 _DEFINE_VEC4_CAST(uint, uchar)
+_DEFINE_VEC4_CAST(short, short)
+_DEFINE_VEC4_CAST(short, ushort)
 _DEFINE_VEC4_CAST(short, char)
 _DEFINE_VEC4_CAST(short, uchar)
+_DEFINE_VEC4_CAST(float, int)
+_DEFINE_VEC4_CAST(float, uint)
 _DEFINE_VEC4_CAST(float, half)
 
 _DEFINE_VEC4_TO_VEC2_CAST(int, int)
-_DEFINE_VEC4_TO_VEC2_CAST(uint, uint)
-_DEFINE_VEC4_TO_VEC2_CAST(float, float)
-_DEFINE_VEC4_TO_VEC2_CAST(short, short)
-_DEFINE_VEC4_TO_VEC2_CAST(short, char)
+_DEFINE_VEC4_TO_VEC2_CAST(int, uint)
 _DEFINE_VEC4_TO_VEC2_CAST(int, short)
 _DEFINE_VEC4_TO_VEC2_CAST(int, char)
+_DEFINE_VEC4_TO_VEC2_CAST(uint, uint)
 _DEFINE_VEC4_TO_VEC2_CAST(uint, ushort)
 _DEFINE_VEC4_TO_VEC2_CAST(uint, uchar)
+_DEFINE_VEC4_TO_VEC2_CAST(short, short)
+_DEFINE_VEC4_TO_VEC2_CAST(short, ushort)
+_DEFINE_VEC4_TO_VEC2_CAST(short, char)
+_DEFINE_VEC4_TO_VEC2_CAST(float, float)
```

Comment:
And again. A whole bunch more of the same type on both sides, going to ignore them from now on, but if they are indeed not needed, please address all the cases.

Seems that this patch adds some more casts, as well as reshuffles the existing ones. In an ideal git scenario, the reshuffling would be in a separate NFC PR, so it's easier to spot what is actually going on.

Reply:
Right, I understand now. Would you be bale to verify that the identity cast is indeed optimised out in the compiler and does not make its way to PTX?

Reply:
I guess it's your call, so I'm happy to go with your judgement, but it would appear to me that for the casts in question they are either:
* optimised out, so there is no problem,
* or are generating the cast instruction in which case the macro should be re-factored?

Reply:
This is the sort of work that never gets done, and there's no point refactoring if you have the opportunity to do it right the first time. If you can verify that the cast doesn't end up with extraneous code in the kernel now, I think that would placate Jakub somewhat. It might "just work" ;)

Reply:
@jchlanda @ldrumm I understand your concerns.

I've looked into this and the output of `cuobjdump` called on my below e2e test is identical regardless of using these macros with the identity cast or an older version of `image.cl` where I don't have this redundancy.

With this, **I believe the identity cast is optimized out.**

Example PTX for the instance of reading 2 floats then writing their sum, where the results read would be cast with `cast_int_to_int()`.

```
suld.b.a1d.b32.clamp {%r10}, [%rd1, {%r8, %r9}];
mov.b32 %f1, %r10;                              
suld.b.a1d.b32.clamp {%r11}, [%rd2, {%r8, %r9}];
mov.b32 %f2, %r11;                              
add.rn.f32 %f3, %f1, %f2;                       
mov.b32 %r12, %f3;                              
sust.b.a1d.b32.clamp [%rd3, {%r8, %r9}], {%r12};
```

Note: This is my first time emitting and analysing PTX from one of the e2e tests, so for sanity check I removed the `as_float()` cast (which is used) and successfully observed this difference in the PTX emitted. That is to say, if you think my conclusion above isn't sufficient then please do share.

Reply:
Great, thank you for checking that, much appreciated.